### ISSUE TITLE
Treating tabs ('\t') as user defined symbols

### ISSUE THREAD

**Issue by pierre-dlsi** [google/sentencepiece#709]
I want to use '_sentencepiece_' to try different tokenization approaches for a music representation language (humdrum **kern). It uses the \<TAB\> character ('\t') as a special kind of delimiter, different from any other whitespace character. In order to encode/decode sentences correctly, \<TAB\>s must be preserved.

As far as I know, sentencepiece treats \<TAB\> as any other whitespace, assigning them the '▁' symbol, replacing <TAB>s by whitespaces in the decoded output. I tried using '_spm_train_' with tabs as a user defined symbol, as follows:

```
$ spm_train --input kern_text.krn --model_prefix=kern --user_defined_symbols="$(printf '\011')" 
```

(that's the way I found to specify the \<TAB\> character on the command line)

However, _spm_train_ terminates with an error when trying to save the model:

> sentencepiece_trainer.cc(77) LOG(INFO) Starts training with :
> trainer_spec {
>   input: kern_text.krn
>... 
> trainer_interface.cc(385) LOG(INFO) Loaded all 2057867 sentences
> trainer_interface.cc(400) LOG(INFO) Adding meta_piece: \<unk\>
> trainer_interface.cc(400) LOG(INFO) Adding meta_piece: \<s\>
> trainer_interface.cc(400) LOG(INFO) Adding meta_piece: \</s\>
> trainer_interface.cc(400) LOG(INFO) Adding meta_piece:               <--- \\t should be here
> trainer_interface.cc(405) LOG(INFO) Normalizing sentences...
> ...(everything looks ok)...
> trainer_interface.cc(615) LOG(INFO) Saving model: kern.model
> spm_train_main.cc(258) [_status.ok()] Internal: src/trainer_interface.cc(563) [!sp->piece().empty()]
> Program terminated with an unrecoverable error.

This happens for any model type. Using the `--allow_whitespace_only_pieces` option does not help either. Trying other non-control ASCII characters in the 'printf' command works fine.

I'm not sure if I'm doing it correctly, any advice is appreciated.

**Comment by xiefangqi** [comment]
Hi pierre-dlsi, this problem is from how sentencepiece parsing args, in sentencepiece/src/util.cc, when args contains whitespace or '\t', these special characters will be skiped:
`
std::vector<std::string> StrSplitAsCSV(absl::string_view text) {
  std::string buf = std::string(text);
  char *str = const_cast<char *>(buf.data());
  char *eos = str + text.size();
  char *start = nullptr;
  char *end = nullptr;

  std::vector<std::string> result;
  for (; str < eos; ++str) {
    while (*str == ' ' || *str == '\t') ++str; // here '\t' is skipped
`

**Comment by xiefangqi** [comment]
I think sentencepiece does not support '\t' as user defined symbols yet.

**Comment by pierre-dlsi** [comment]
Thnaks for the naswer, @xiefangqi. 

I did a workaround, preprocessing the input to change '\t' to a string "<T>", so I can define it as a special symbol.

May be I would try to modify the code and recompile it, and see if it does the trick.

Thanks again.

**Comment by taku910** [comment]
Fixed this issue in version 0.1.97.   We can now specify '\t' as user defined symbols.
